### PR TITLE
Monitor processes, unsubscribe them when they go down

### PR DESCRIPTION
**Pull request by acco**
* Use Process.monitor to monitor recipients
* Handle DOWN messages from recipients. Unsub them from NATS, remove from state.
* Process.demonitor whenever a receipient usubs

Addresses https://github.com/nats-io/nats.ex/issues/33

Let me know if this looks good, doing some testing on it now!

### CODE REVIEW

**Comment by acco** [comment]
Hey @mmmries – thanks for your work on this project! Just an update: been running this change in prod since I opened the PR and it seems to do the trick

**Comment by mmmries** [comment]
Hey @acco thanks so much for the submissions 🙌 

I'm thinking we should ask @autodidaddict to check this out as well as he's been helping out with the Elixir client and has a lot more context from within Synadia. Did you run any benchmarks on this by any chance? I was going back and forth on whether it would be best to handle exceptions when attempting to deliver messages vs actively monitoring client processes.

**Comment by acco** [comment]
Hey @mmmries – my pleasure!

I have not run any benchmarks on this.

File: lib/gnat.ex [hunk 1]
```diff
@@ -493,12 +494,25 @@ defmodule Gnat do
...
   end
 
+  defp unsub_sid(sid, opts, state) do
+    case Map.get(state.receivers, sid) do
+      nil ->
+        state
+    %{monitor_ref: ref} ->
+      command = Command.build(:unsub, sid, opts)
+      :ok = socket_write(state, command)
+      Process.demonitor(ref)
```
Comment (autodidaddict) [inline lib/gnat.ex:504]: I had no idea that demonitor was idempotent. Nice!

**Review by autodidaddict — APPROVED** [review]
LGTM. I can't see anything in here that would affect performance. It looks like the worst it's doing is adding an `Enum.find` during a process death, which is acceptable

**Comment by mmmries** [comment]
I ran a few benchmarks on this and found that things all looked good. I ran the `bench/request.exs` single-core performance test that does a roundtrip call and I also ran the `bench/server.exs` + `bench/client.exs` combined which goes between multiple VM instances with lots of parallelism. I ran this with 128 requesters and 5000 messages per requester

### main branch

```
mries@mries-ltmtqyn nats.ex % mix run bench/client.exs
It took 8.465676sec
	75599.39690581118 req/sec
	2362.4811533065995 kb/sec
	min: 103µs
	max: 15433µs
	median: 1444µs
	90th percentile: 2402µs
	average: 1575µs
	103,15433,1444,2402,1575,75599.39690581118,2362.4811533065995
```

### This PR branch

```
mries@mries-ltmtqyn nats.ex % mix run bench/client.exs
It took 8.452638sec
	75716.00723939674 req/sec
	2366.125226231148 kb/sec
	min: 92µs
	max: 14753µs
	median: 1440µs
	90th percentile: 2415µs
	average: 1575µs
	92,14753,1440,2415,1575,75716.00723939674,2366.125226231148
```

Ran this on erlang 25.1 and elixir 1.14.0 on a macbook pro. Seeing no significant different in performance which is amazing. I ❤️💙💜💚💛🧡🤎🖤 the Erlang VM for giving us these kinds of powerful tools

**Comment by mmmries** [comment]
Released as `gnat` version `1.8.1` https://hex.pm/packages/gnat/1.8.1

Thanks @acco for the contribution

**Comment by autodidaddict** [comment]
Can we get a github release for this?

**Comment by mmmries** [comment]
Yes, thanks for the reminder @autodidaddict

I created release 1.8.1 in github which also contains all of the jetstream integration work you did in version 1.8

**Comment by acco** [comment]
Awesome, thanks all! (So cool to see that benchmark)

**Pull request merged**